test/integration/src/test/java/io/pravega/test/integration/MetricsTest.java

```diff
@@ -206,7 +206,10 @@ public void metricsTimeBasedCacheEvictionTest() throws Exception {
             AssertExtensions.assertEventuallyEquals(bytesWritten, () -> {
                 return MetricRegistryUtils.getCounter(SEGMENT_READ_BYTES, streamTags) == null
                         ? -1 : (long) MetricRegistryUtils.getCounter(SEGMENT_READ_BYTES, streamTags).count();
-            }, 10000);
+            }, 2000);
+
+            //Wait for cache eviction to happen
+            Thread.sleep(5000);
```

Comment:
Is there really no other way of simulating the passage of time? Can we not have an internal method that forces a cache eviction?

Reply:
No easy way at least; and this is integration test so let's do it "naturally".

Reply:
Relying upon sleeps in tests is bad practice for at least two reasons:

- It increases the total length of the build often unnecessarily
- Due to the non-determinism of process and thread scheduling, it leads to false positives. 

Please find another way to determine that the element has been evicted from the cache.

Reply:
Reduced the eviction time to 2 seconds, and reduced the sleep time to 4 seconds. The total test execution time is dropped from 54 sec to 24 sec.
Note this is a "time based" integration test trying to ensure metric cache indeed evicts as expected from an end-to-end integration perspective. The unit tests of Cache evictions are already covered by SegmentStatsRecorderTest.java, which uses a TestContext as the run time, and supplies a mocked cache.
If we manipulate the cache too much in this integration test, then it's overlapping the unit test SegmentStatsRecorderTest; more importantly, it's losing the point of integration test.

Reply:
Just to add one more comment - I spent quite some time trying to find a way to manipulate the cache eviction so we can have more precise control of the test execution. Firstly it is near impossible to manipulate internal components of Pravega in an integration test environment; secondly when I tried it further, this test became more and more like SegmentStatsRecorderTest, a pure unit test, and starts to lose the point of integration test.